### DISCUSSION TITLE
How to stop on ongoing motor motion using another condition?

### DISCUSSION THREAD

**Discussion by laurensvalk** [pybricks/support#1621] [Q&A] [answered]
This was asked via email, so I've opened a discussion here so everyone can see the possible answer, and share other solutions.

> A particularly valuable case is when the Lego robot encounters an obstacle and cannot complete the programmed motor movement (motion or arm movement), how to handle such a situation programmatically?
> 
> How to detect that the robot has caught something while driving and cannot continue moving?
> 
> These are typical situations, detection of which and response to them would make the robot more "intelligent".
> 
> I will be grateful for your tips.

As a concrete example, let's say we want to rotate motor A for 3600 degrees (10 rotations), but stop earlier if we see an obstacle or the motor gets stuck, so we don't keep waiting forever.

**Comment by laurensvalk** [comment, marked as answer]
Another way of phrasing this more systematically, is to say we want to rotate the motor
- until it reaches 3600 degrees; or
- until it gets stuck/stalled; or
- until the sensor sees an obstacle...

_... whichever comes first_.

There are a few solutions.

**The easy way with Pybricks**

You can implement this quite literally with the multitask block, and set it to complete _when one task finishes_:

![multitask](https://github.com/pybricks/support/assets/12326241/4fb40f5b-93ec-41a5-9a41-dfc35046e67a)

It will do three tasks at once: rotate, wait for stall, and wait for sensor. It completes when one of them completes, which is exactly what we wanted!

Then the program continues with the rest, which is just a sound in this case.

**The hard way 1 (synchronously using classic apps/solutions)**

In the classic solution, you could:

- start the motor to run forever
- while not (3600 degrees turned OR stalled OR object detected):
    -  wait and do nothing
- stop the motor
- use nested if/elsif/else to find out _which_ of the three conditions caused the stop
- based on that, decide what to do next

**The hard way 2 (asynchronously using classic apps/solutions)**

- start the motor on one stack
- use event hats to detect objects or getting stuck
- stop the motor from the other stack
- wait for broadcast on the main stack
- depending on which broadcast was received, decide what to do.
- reset all state/mutex/lock variables so you could repeat all this again

**Reply by trwhitt** [reply]
I have a quick question that maybe you can help me with as we dabble with Pybricks.  How would one do this inside of another task?  We keep getting:

TypeError: 'int' object isn't iterable

Attached is a picture, thank you!

<img width="592" alt="Screenshot 2024-11-01 at 7 27 37 PM" src="https://github.com/user-attachments/assets/f0cfa95c-4669-48ff-9c05-193a678beefd">

**Reply by laurensvalk** [reply]
This snippet looks fine, so I suspect the issue is coming from somewhere else. Could you share the generated code and the full error so we can see which line is causing it?

**Reply by trwhitt** [reply]
Laurensvalk! 
Thanks for replying.  We tried some things this weekend but still couldn't solve it yet so we kept working while avoiding multitasking.  We'd really like to utilize some stall detection!

<img width="373" alt="Screenshot 2024-11-04 at 8 09 36 AM" src="https://github.com/user-attachments/assets/1676f29c-8b27-4ab8-9436-d525c643f16f">

<img width="845" alt="Screenshot 2024-11-04 at 8 14 06 AM" src="https://github.com/user-attachments/assets/ce54b1cd-0c27-48f6-91f1-90a9a8478b78">

<img width="523" alt="Screenshot 2024-11-04 at 8 14 34 AM" src="https://github.com/user-attachments/assets/28bb302d-d1ff-4c14-9d75-ff6b6e6cdf10">

**Reply by laurensvalk** [reply]
I'm going to guess that `prior_program` and `next_program` are functions that you write yourself, perhaps in another file?

If you use multitasking, it is best to do so everywhere. Since it is being called as `await prior_program`, it is expecting a function of the form `async def` instead of just `def`.

**Comment by amberneb** [comment]
Thank you! We will give this a try!

**Comment by bcfletcher** [comment]
so something like this in Python:

```python
from pybricks.hubs import PrimeHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch, multitask, run_task, hub_menu
from umath import sin, pi

hub = PrimeHub()

topAttachment = Motor(Port.D,Direction.CLOCKWISE,gears=[20,28],reset_angle=True)

async def isStalled() -> bool :
    print("isStalled()",topAttachment.stalled())
    while topAttachment.stalled() == False: 
        await wait(1)
    print("stalled")
    return topAttachment.stalled()

async def moveArm(speed,angle) -> bool:
    print ("moveArm() speed=",speed," angle=",angle)
    await topAttachment.run_target(speed,angle,then=Stop.HOLD)
    return True

async def main():

    print("topAttachment Angle=",topAttachment.angle())
    topAttachment.reset_angle(0)

    print("start")
    await multitask(moveArm(50,180),isStalled(),race=True)
    await multitask(moveArm(50,-180),isStalled(),race=True)
    print ("end")

run_task(main())
```
